codegen/projections/high_score_service/lib/high_score_service/endpoint.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+# WARNING ABOUT GENERATED CODE
+#
+# This file was code generated using smithy-ruby.
+# https://github.com/awslabs/smithy-ruby
+#
+# WARNING ABOUT GENERATED CODE
+
+module HighScoreService
+  module Endpoint
+    Params = ::Struct.new(
+      :endpoint,
+      keyword_init: true
+    ) do
+      include Hearth::Structure
+    end
+
+    class Provider
+      def resolve_endpoint(params)
+        endpoint = params.endpoint
+
+        if (endpoint != nil)
+          return Hearth::RulesEngine::Endpoint.new(uri: endpoint)
+        end
+        raise ArgumentError, "Endpoint is not set - you must configure an endpoint."
+
+        raise ArgumentError, 'No endpoint could be resolved'
```

Comment:
Yeah - in this case the defined rules have an error rule that will always be hit.  Generically however, they do not and the Smithy [specification](https://smithy.io/2.0/additional-specs/rules-engine/specification.html) on the rules engine has:

> If the list of rules is exhausted with none evaluated successfully, then an implementation of the rule set engine MUST stop and return an error indicating that rule exhaustion has occurred. Rule authors SHOULD use either an [endpoint rule](https://smithy.io/2.0/additional-specs/rules-engine/specification.html#rules-engine-endpoint-rule-set-endpoint-rule) or [error rules,](https://smithy.io/2.0/additional-specs/rules-engine/specification.html#rules-engine-endpoint-rule-set-error-rule) with an empty set of conditions to provide a more meaningful default or error depending on the scenario.

There are a few ways we could approach this:
1. [Current approach] Always add the default error case last in codegen.  It may be unreachable.
2. Attempt to detect if the ruleset always has a terminal case.  (Generically this is difficult but technically do-able)
3. Always add a default rule to endpoint ruleset.  Then codegen will take care of adding a case for it.  It may be unreachable.

We could also have the Endpoint middleware handle this case - if it returns nil, raise an exception.  However, I think the default error behavior should probably be part of the Endpoint Provider.

Reply:
For now I've gone ahead with a variation of 2 above - specifically if the last rule does not have any conditions, then we know for sure we do not need the fallback error (Because that rule is guaranteed to be hit).   This prevents unreachable fallback error in a lot of cases, but of course not generically (there are complex rulesets that will always reach a leaf, but which don't have a last rule w. empty conditions).